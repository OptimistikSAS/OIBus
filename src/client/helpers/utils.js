const dynamicSort = (property) => {
  let prop = property
  let sortOrder = 1
  if (prop[0] === '-') {
    sortOrder = -1
    prop = prop.substr(1)
  }
  return (a, b) => {
    let result
    const valueA = a[prop].toString()
    const valueB = b[prop].toString()
    if (valueA < valueB) {
      result = -1
    } else if (valueA > valueB) {
      result = 1
    } else {
      result = 0
    }
    return result * sortOrder
  }
}

export default { dynamicSort }
