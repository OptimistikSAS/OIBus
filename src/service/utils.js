const fs = require('node:fs/promises')
const { createReadStream, createWriteStream } = require('node:fs')
const zlib = require('node:zlib')
const path = require('node:path')

const minimist = require('minimist')
const { DateTime } = require('luxon')

const COMPRESSION_LEVEL = 9

/**
 * Get config file from console arguments
 * @returns {Object} - the config file and check argument
 */
const getCommandLineArguments = () => {
  const args = minimist(process.argv.slice(2))
  const { config = './oibus.json', check = false } = args
  return { configFile: path.resolve(config), check }
}

/**
 * Method to return a delayed promise.
 * @param {Number} timeout - The delay
 * @return {Promise<any>} - The delay promise
 */
const delay = async (timeout) => new Promise((resolve) => {
  setTimeout(resolve, timeout)
})

/**
 * Compute a list of end interval from a startTime, endTime and maxInterval.
 * @param {Date} startTime - The start of the interval
 * @param {Date} endTime - The end of the interval
 * @param {Number} maxInterval - The interval in s to split the [startTime - endTime] interval into
 * @returns {[{Date, Date}]} - The list of intervals
 */
const generateIntervals = (startTime, endTime, maxInterval) => {
  const originalInterval = endTime.getTime() - startTime.getTime()
  if (maxInterval > 0 && (endTime.getTime() - startTime.getTime()) > 1000 * maxInterval) {
    const numberOfInterval = originalInterval / (maxInterval * 1000)
    const intervalLists = []
    for (let i = 0; i < numberOfInterval; i += 1) {
      // Compute the newStartTime and the newEndTime for each interval
      const newStartTime = new Date(startTime.getTime() + i * 1000 * maxInterval)
      const newEndTime = new Date(startTime.getTime() + (i + 1) * 1000 * maxInterval)

      // If the newEndTime is bigger than the original endTime, the definitive end of the interval must be endTime
      intervalLists.push({
        startTime: newStartTime,
        endTime: newEndTime < endTime ? newEndTime : endTime,
      })
    }
    return intervalLists
  }
  return [{ startTime, endTime }]
}

/**
 * Create a folder if it does not exist
 * @param {String} folder - The folder to create
 * @return {Promise<void>} - The result promise
 */
const createFolder = async (folder) => {
  const folderPath = path.resolve(folder)
  try {
    await fs.stat(folderPath)
  } catch (error) {
    await fs.mkdir(folderPath, { recursive: true })
  }
}

/**
 * Replace the variables such as @CurrentDate in the file name with their values
 * @param {String} filename - The filename to change
 * @param {Number} queryPart - The part of the query being executed
 * @param {String} connectorName - The name of the connector
 * @return {String} - The renamed file name
 */
const replaceFilenameWithVariable = (filename, queryPart, connectorName) => filename.replace('@CurrentDate', DateTime.local()
  .toFormat('yyyy_MM_dd_HH_mm_ss_SSS'))
  .replace('@ConnectorName', connectorName)
  .replace('@QueryPart', `${queryPart}`)

/**
 * Generate date based on the configured format taking into account the timezone configuration.
 * Ex: With timezone "Europe/Paris" the date "2019-01-01 00:00:00" will be converted to "Tue Jan 01 2019 00:00:00 GMT+0100"
 * @param {String} date - The date to parse and format
 * @param {String} timezone - The timezone to use to replace the timezone of the date
 * @param {String} dateFormat - The format of the date
 * @returns {String} - The formatted date with timezone
 */
const generateDateWithTimezone = (date, timezone, dateFormat) => {
  if (DateTime.fromISO(date).isValid) {
    return date
  }
  return DateTime.fromFormat(date, dateFormat, { zone: timezone }).toJSDate().toISOString()
}

/**
 * Compress the specified file
 * @param {String} input - The path of the file to compress
 * @param {String} output - The path to the compressed file
 * @returns {Promise<void>} - The result promise
 */
const compress = async (input, output) => new Promise((resolve, reject) => {
  const readStream = createReadStream(input)
  const writeStream = createWriteStream(output)
  const gzip = zlib.createGzip({ level: COMPRESSION_LEVEL })
  readStream
    .pipe(gzip)
    .pipe(writeStream)
    .on('error', (error) => {
      reject(error)
    })
    .on('finish', () => {
      resolve()
    })
})

/**
 * Check if a file exists in async way
 * @param {String} filePath - The file path to test
 * @returns {Promise<boolean>} - If the file exists or not
 */
const filesExists = async (filePath) => {
  try {
    await fs.stat(filePath)
  } catch {
    return false
  }
  return true
}

module.exports = {
  getCommandLineArguments,
  delay,
  generateIntervals,
  createFolder,
  replaceFilenameWithVariable,
  generateDateWithTimezone,
  compress,
  filesExists,
}
