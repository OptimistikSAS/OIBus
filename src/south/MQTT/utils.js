const mqttWildcard = require('mqtt-wildcard')
const { vsprintf } = require('sprintf-js')
const { generateDateWithTimezone } = require('../../services/utils')

/**
 * Get pointId.
 * @param {string} topic - The topic
 * @param {object} currentData - The data being parsed
 * @param {string} pointIdPath - The pointId path
 * @param {array} topicList - The list of all topics
 * @return {string | null} - The pointId
 */
const getPointId = (topic, currentData, pointIdPath, topicList) => {
  if (pointIdPath) { // if the pointId is in the data
    if (!currentData[pointIdPath]) {
      throw new Error(`Could node find pointId in path ${pointIdPath} for data: ${JSON.stringify(currentData)}`)
    }
    return currentData[pointIdPath]
  }

  // else, the pointId is in the topic
  let pointId = null
  const matchedPoints = []

  topicList.forEach((point) => {
    const matchList = mqttWildcard(topic, point.topic)
    if (Array.isArray(matchList)) {
      if (!pointId) {
        const nrWildcards = (point.pointId.match(/[+#]/g) || []).length
        if (nrWildcards === matchList.length) {
          const normalizedPointId = point.pointId.replace(/[+#]/g, '%s')
          pointId = vsprintf(normalizedPointId, matchList)
          matchedPoints.push(point)
        } else {
          throw new Error(`Invalid point configuration: ${JSON.stringify(point)}`)
        }
      } else {
        matchedPoints.push(point)
      }
    }
  })

  if (matchedPoints.length > 1) {
    throw new Error(`${topic} should be subscribed only once but it has the
     following subscriptions: ${JSON.stringify(matchedPoints)}`)
  }

  return pointId
}

/**
 * Get timestamp.
 * @param {string} elementTimestamp - The element timestamp
 * @param {string} timestampOrigin - The timestamp origin
 * @param {string} timestampFormat - The timestamp format
 * @param {string} timezone - The timezone
 * @return {string} - The formatted timestamp
 */
const getTimestamp = (elementTimestamp, timestampOrigin, timestampFormat, timezone) => {
  let timestamp = new Date().toISOString()

  if (timestampOrigin === 'payload') {
    if (timezone && elementTimestamp) {
      timestamp = generateDateWithTimezone(elementTimestamp, timezone, timestampFormat)
    }
  }
  return timestamp
}

/**
 *
 * @param {object} data - The data to format
 * @param {string} topic - The mqtt topic
 * @param {object} formatOptions - The formatting options
 * @param {array} topicList - The list of all topics
 * @returns {{pointId: string, data: {value: *, quality: *}, timestamp: string}|null} - the formatted data
 */
const formatValue = (data, topic, formatOptions, topicList) => {
  const {
    timestampPath,
    timestampOrigin,
    timestampFormat,
    timezone,
    valuePath,
    pointIdPath,
    qualityPath,
  } = formatOptions
  const dataPointId = getPointId(topic, data, pointIdPath, topicList)
  if (dataPointId) {
    const dataTimestamp = getTimestamp(
      data[timestampPath],
      timestampOrigin,
      timestampFormat,
      timezone,
    )
    const dataValue = data[valuePath]
    const dataQuality = data[qualityPath]
    // delete fields to avoid duplicates in the returned object
    delete data[timestampPath]
    delete data[valuePath]
    delete data[pointIdPath]
    delete data[qualityPath]
    return {
      pointId: dataPointId,
      timestamp: dataTimestamp,
      data: {
        ...data,
        value: dataValue,
        quality: dataQuality,
      },
    }
  }
  throw new Error('PointId can\'t be determined. '
        + `The following value ${JSON.stringify(data)} is not saved.`)
}

module.exports = { formatValue }
