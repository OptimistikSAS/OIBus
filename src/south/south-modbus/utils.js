/**
 * Retrieve the number of Words in Modbus associated to each data type
 * @param {String} dataType - The data type in the buffer
 * @returns {Number} - The number of words to request
 */
const getNumberOfWords = (dataType) => {
  if (['Int16', 'UInt16'].includes(dataType)) {
    return 1
  } if (['UInt32', 'Int32', 'Float'].includes(dataType)) {
    return 2
  } if (['BigUInt64', 'BigInt64', 'Double'].includes(dataType)) {
    return 4
  }
  return 1
}

export default getNumberOfWords
