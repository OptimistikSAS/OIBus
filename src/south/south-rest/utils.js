import https from 'node:https'
import http from 'node:http'

import csv from 'papaparse'

import oiaTimeValues from './formatters/oia-time-values.js'
import slims from './formatters/slims.js'

const parsers = {
  Raw: (httpResults) => ({ httpResults, latestDateRetrieved: new Date().toISOString() }),
  'OIAnalytics time values': oiaTimeValues,
  SLIMS: slims,
}

/**
 * Some API such as SLIMS uses a body with GET. It's not standard and requires a specific implementation
 * @param {String} body - The body to append to the HTTP GET query
 * @param {Object} options - HTTP options
 * @returns {Promise<Object>} - The HTTP response
 */
const httpGetWithBody = (body, options) => new Promise((resolve, reject) => {
  const callback = (response) => {
    let str = ''
    response.on('data', (chunk) => {
      str += chunk
    })
    response.on('end', () => {
      try {
        const parsedResult = JSON.parse(str)
        resolve(parsedResult)
      } catch (error) {
        reject(error)
      }
    })
  }

  const req = (options.protocol === 'https:' ? https : http).request(options, callback)
  req.on('error', (e) => {
    reject(e)
  })
  req.write(body)
  req.end()
})

/**
 * @param {Date} startTime - The start time
 * @param {Date} endTime - The end time
 * @param {Object[]} queryParams - The list of query params to format with variables
 * @param {'ISO'|'number'} variableDateFormat - How to format the date variables in the query params
 * @returns {String} - The query params in string format with variables' value
 */
const formatQueryParams = (startTime, endTime, queryParams, variableDateFormat) => {
  if (!queryParams || queryParams.length === 0) {
    return ''
  }
  let queryParamsString = '?'
  queryParams.forEach((queryParam, index) => {
    let value
    switch (queryParam.queryParamValue) {
      case '@StartTime':
        value = variableDateFormat === 'ISO'
          ? new Date(startTime).toISOString() : new Date(startTime).getTime()
        break
      case '@EndTime':
        value = variableDateFormat === 'ISO'
          ? new Date(endTime).toISOString() : new Date(endTime).getTime()
        break
      default:
        value = queryParam.queryParamValue
    }
    queryParamsString += `${encodeURIComponent(queryParam.queryParamKey)}=${encodeURIComponent(value)}`
    if (index < queryParams.length - 1) {
      queryParamsString += '&'
    }
  })
  return queryParamsString
}

/**
 * Generate CSV file from the values.
 * @param {Object[]} results - The HTTP query result
 * @param {String} delimiter - The delimiter to use for the CSV
 * @returns {String} - The CSV content
 */
const generateCSV = (results, delimiter) => {
  const options = {
    header: true,
    delimiter,
  }
  return csv.unparse(results, options)
}

export { parsers, httpGetWithBody, formatQueryParams, generateCSV }
