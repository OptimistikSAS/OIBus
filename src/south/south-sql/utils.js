import csv from 'papaparse'
import { DateTime } from 'luxon'

/**
 * Format date taking into account the timezone configuration.
 * Since we don't know how the date is actually stored in the database, we read it as UTC time
 * and format it as it would be in the configured timezone.
 * Ex: With timezone "Europe/Paris" the date "2019-01-01 00:00:00" will be converted to "Tue Jan 01 2019 00:00:00 GMT+0100"
 * @param {Date} date - The date to format
 * @param {String} timezone - The timezone to use to replace the timezone of the date
 * @param {String} dateFormat - The format of the date to use for the return result
 * @returns {String} - The formatted date with timezone
 */
const formatDateWithTimezone = (date, timezone, dateFormat) => DateTime.fromJSDate(date, { zone: 'utc' }).toFormat(dateFormat)

/**
 * Generate CSV file from the values.
 * @param {Object[]} result - The query result
 * @param {String} timezone - The timezone of the database
 * @param {String} dateFormat - The date format for the date fields
 * @param {String} delimiter - The delimiter to use for the CSV
 * @returns {Promise<String>} - The CSV content
 */
const generateCSV = (result, timezone, dateFormat, delimiter) => {
  // loop through each value and format date to timezone if value is Date
  result.forEach((row) => {
    Object.keys(row).forEach((key) => {
      const value = row[key]
      if (value instanceof Date) {
        row[key] = formatDateWithTimezone(value, timezone, dateFormat)
      }
    })
  })
  const options = {
    header: true,
    delimiter,
  }
  return csv.unparse(result, options)
}

/**
 * Parse an entry list and get the most recent date
 * @param {Object[]} entryList - The SQL results
 * @param {Date} startTime - The reference date
 * @param {String} timeColumn - The time column to look for in the entry list
 * @param {String} timezone - The timezone of the database
 * @return {Date} - The most recent date
 */
const getMostRecentDate = (entryList, startTime, timeColumn, timezone) => {
  let newLastCompletedAt = startTime
  entryList.forEach((entry) => {
    if (entry[timeColumn]) {
      let entryDate
      if (entry[timeColumn] instanceof Date) {
        entryDate = entry[timeColumn]
      } else if (typeof entry[timeColumn] === 'number') {
        entryDate = DateTime.fromMillis(entry[timeColumn], { zone: timezone })
          .setZone('utc')
          .toJSDate()
      } else if (DateTime.fromISO(entry[timeColumn], { zone: timezone }).isValid) {
        entryDate = DateTime.fromISO(entry[timeColumn], { zone: timezone })
          .setZone('utc')
          .toJSDate()
      } else if (DateTime.fromSQL(entry[timeColumn], { zone: timezone }).isValid) {
        entryDate = DateTime.fromSQL(entry[timeColumn], { zone: timezone })
          .setZone('utc')
          .toJSDate()
      }
      if (entryDate > new Date(newLastCompletedAt)) {
        newLastCompletedAt = DateTime.fromMillis(entryDate.setMilliseconds(entryDate.getMilliseconds() + 1))
          .setZone('utc')
          .toJSDate()
      }
    }
  })
  return newLastCompletedAt
}

/**
 * Get all occurrences of a substring with a value
 * @param {String} str - The string to look for occurrences in
 * @param {String} keyword - The keyword
 * @param {Object} value - The value to assign to the occurrences index
 * @return {Object[]} - The result as { index, value}
 */
const getOccurrences = (str, keyword, value) => {
  const occurrences = []
  let occurrenceIndex = str.indexOf(keyword, 0)
  while (occurrenceIndex > -1) {
    occurrences.push({
      index: occurrenceIndex,
      value,
    })
    occurrenceIndex = str.indexOf(keyword, occurrenceIndex + 1)
  }
  return occurrences
}

/**
 * Generate replacements parameters
 * @param {String} query - The query
 * @param {Date} startTime - The StartTime
 * @param {Date} endTime - The EndTime
 * @return {Object[]} - The replacement parameters
 */
const generateReplacementParameters = (query, startTime, endTime) => {
  const startTimeOccurrences = getOccurrences(query, '@StartTime', startTime)
  const endTimeOccurrences = getOccurrences(query, '@EndTime', endTime)
  const occurrences = startTimeOccurrences.concat(endTimeOccurrences)
  occurrences.sort((a, b) => (a.index - b.index))
  return occurrences.map((occurrence) => occurrence.value)
}

export { generateCSV, getMostRecentDate, generateReplacementParameters }
