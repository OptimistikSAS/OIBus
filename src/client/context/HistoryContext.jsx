import React from 'react'
import PropTypes from 'prop-types'
import objectPath from 'object-path'
import apis from '../services/apis'
import utils from '../helpers/utils'
import { ConfigContext } from './ConfigContext.jsx'

const reducer = (state, action) => {
  const {
    name,
    value,
    config,
    type,
    validity,
  } = action
  const newState = utils.jsonCopy(state)
  switch (type) {
    case 'reset':
      return utils.jsonCopy(config)
    case 'update':
      if (!newState.errors) newState.errors = {}
      if (validity) {
        newState.errors[name] = validity
      } else {
        delete newState.errors[name]
        if (Object.keys(newState.errors).length === 0) delete newState.errors
      }
      objectPath.set(newState, name, value)
      return newState
    case 'deleteRow': {
      if (newState.errors) {
        const keys = Object.keys(newState.errors)
          .filter((key) => key.startsWith(name))
        keys.forEach((key) => delete newState.errors[key])
        if (Object.keys(newState.errors).length === 0) delete newState.errors
      }
      objectPath.del(newState, name)
      return newState
    }
    case 'deleteAllRows':
      objectPath.empty(newState, name)
      return newState
    case 'addRow':
      objectPath.push(newState, name, value)
      return newState
    case 'importPoints':
      objectPath.set(newState, name, value)
      return newState
    default:
      throw new Error(`unknown action type: ${type}`)
  }
}
const historyConfigInitialState = []
const HistoryConfigContext = React.createContext(historyConfigInitialState)
const HistoryConfigProvider = ({ children }) => {
  const [newHistoryConfig, dispatchNewHistoryConfig] = React.useReducer(reducer, historyConfigInitialState)
  const [activeHistoryConfig, setActiveHistoryConfig] = React.useState([])
  const {
    protocolList,
    apiList,
  } = React.useContext(ConfigContext)

  React.useEffect(() => {
    let mounted = true
    const fetchActiveHistoryConfig = async () => {
      try {
        const resp = await apis.getHistoryConfig()
        if (mounted && resp) {
          dispatchNewHistoryConfig({
            type: 'reset',
            config: resp.config,
          })
          setActiveHistoryConfig(utils.jsonCopy(resp.config))
        }
      } catch (error) {
        console.error(error)
      }
    }
    fetchActiveHistoryConfig()
    return () => {
      mounted = false
    }
  }, [])

  const configValueProvided = React.useMemo(
    () => ({
      newHistoryConfig,
      dispatchNewHistoryConfig,
      activeHistoryConfig,
      setActiveHistoryConfig,
      apiList,
      protocolList,
    }),
    [newHistoryConfig,
      dispatchNewHistoryConfig,
      activeHistoryConfig,
      setActiveHistoryConfig,
      apiList,
      protocolList],
  )
  return (
    <HistoryConfigContext.Provider
      value={configValueProvided}
    >
      {children}
    </HistoryConfigContext.Provider>
  )
}

HistoryConfigProvider.propTypes = { children: PropTypes.element.isRequired }
export { HistoryConfigContext, HistoryConfigProvider, reducer }
