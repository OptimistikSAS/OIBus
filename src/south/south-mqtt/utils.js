import mqttWildcard from 'mqtt-wildcard'
import { vsprintf } from 'sprintf-js'
import { generateDateWithTimezone } from '../../service/utils.js'

/**
 * Get pointId.
 * @param {String} topic - The topic
 * @param {Object} currentData - The data being parsed
 * @param {String} pointIdPath - The pointId path
 * @param {Object[]} pointsList - The list of all points
 * @return {String | null} - The pointId
 */
const getPointId = (topic, currentData, pointIdPath, pointsList) => {
  if (pointIdPath) { // if the pointId is in the data
    if (!currentData[pointIdPath]) {
      throw new Error(`Could node find pointId in path "${pointIdPath}" for data: ${JSON.stringify(currentData)}`)
    }
    return currentData[pointIdPath]
  }

  // else, the pointId is in the topic
  let pointId = null
  const matchedPoints = []

  pointsList.forEach((point) => {
    const matchList = mqttWildcard(topic, point.topic)
    if (Array.isArray(matchList)) {
      if (!pointId) {
        const nrWildcards = (point.pointId.match(/[+#]/g) || []).length
        if (nrWildcards === matchList.length) {
          const normalizedPointId = point.pointId.replace(/[+#]/g, '%s')
          pointId = vsprintf(normalizedPointId, matchList)
          matchedPoints.push(point)
        } else {
          throw new Error(`Invalid point configuration: ${JSON.stringify(point)}`)
        }
      } else {
        matchedPoints.push(point)
      }
    }
  })

  if (matchedPoints.length > 1) {
    throw new Error(`Topic "${topic}" should be subscribed only once but it has the `
     + `following subscriptions: ${JSON.stringify(matchedPoints)}`)
  } else if (!pointId) {
    throw new Error(`PointId can't be determined. The following value ${JSON.stringify(currentData)} is not saved.`)
  }

  return pointId
}

/**
 * Get timestamp.
 * @param {String} elementTimestamp - The element timestamp
 * @param {'payload'|'oibus'} timestampOrigin - The timestamp origin
 * @param {String} timestampFormat - The timestamp format
 * @param {String} timezone - The timezone
 * @return {String} - The formatted timestamp
 */
const getTimestamp = (elementTimestamp, timestampOrigin, timestampFormat, timezone) => {
  let timestamp = new Date().toISOString()

  if (timestampOrigin === 'payload') {
    if (timezone && elementTimestamp) {
      timestamp = generateDateWithTimezone(elementTimestamp, timezone, timestampFormat)
    }
  }
  return timestamp
}

/**
 *
 * @param {Object} data - The data to format
 * @param {String} topic - The mqtt topic received. It can differ from the pointsList topic which can have a wildcard
 * @param {Object} formatOptions - The formatting options
 * @param {Object[]} pointsList - The list of all topics
 * @returns {{pointId: String, data: {value: any, quality: any}, timestamp: String}|null} - the formatted data
 */
const formatValue = (data, topic, formatOptions, pointsList) => {
  const {
    timestampPath,
    timestampOrigin,
    timestampFormat,
    timezone,
    valuePath,
    pointIdPath,
    qualityPath,
  } = formatOptions
  const dataPointId = getPointId(topic, data, pointIdPath, pointsList)

  const dataTimestamp = getTimestamp(
    data[timestampPath],
    timestampOrigin,
    timestampFormat,
    timezone,
  )
  const dataValue = data[valuePath]
  const dataQuality = data[qualityPath]
  // delete fields to avoid duplicates in the returned object
  delete data[timestampPath]
  delete data[valuePath]
  delete data[pointIdPath]
  delete data[qualityPath]
  return {
    pointId: dataPointId,
    timestamp: dataTimestamp,
    data: {
      ...data,
      value: dataValue,
      quality: dataQuality,
    },
  }
}

export default formatValue
