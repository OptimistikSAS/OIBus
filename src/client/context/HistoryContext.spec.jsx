/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import utils from '../helpers/utils'
import testConfig from '../../../tests/testHistoryConfig'
import { HistoryConfigProvider, reducer } from './HistoryContext.jsx'

// mock fetch
global.fetch = jest.fn().mockImplementation((uri) => {
  let jsonString
  switch (uri) {
    case '/history-query/config':
      jsonString = JSON.stringify({ config: testConfig })
      break
    default:
      jsonString = '""'
      break
  }
  return {
    status: 200,
    text: jest.fn().mockImplementation(() => jsonString),
  }
})

// mock states
const initState = 'activeHistoryConfig'
const setActiveHistoryConfig = jest.fn()
const setState = jest.fn()
React.useState = jest.fn().mockImplementation((init) => {
  if (init?.length === 0 && initState === 'activeHistoryConfig') {
    return [init, setActiveHistoryConfig]
  }
  return [init, setState]
})
// mock dispatchNewConfig
React.useReducer = jest.fn().mockImplementation(() => [testConfig, jest.fn()])
const dispatchNewHistoryConfig = (action) => reducer(testConfig, action)

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('HistoryConfigProvider', () => {
  test('check HistoryConfigProvider', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
  test('check initial data setup', async () => {
    await act(async () => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    expect(setActiveHistoryConfig).toBeCalledWith(testConfig)
    expect(container).toMatchSnapshot()
  })
  test('check initial data fail', async () => {
    const originalError = console.error
    console.error = jest.fn()
    const originalGlobalFetchMock = global.fetch
    global.fetch = jest.fn().mockImplementation(() => ({ status: 500 }))
    await act(async () => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    expect(console.error).toBeCalled()
    global.fetch = originalGlobalFetchMock
    console.error = originalError
  })
  test('check reset action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = { type: 'reset', config: testConfig }
    const newState = dispatchNewHistoryConfig(action)
    expect(newState).toEqual(testConfig)
  })
  test('check update action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = { type: 'update', name: '0.compress', value: false }
    const newState = dispatchNewHistoryConfig(action)
    expect(newState[0].compress).toEqual(false)
  })
  test('check update action with previous error', () => {
    // add some previuous error to test error cleaning
    testConfig.errors = { '0.compress': 'some previous error' }

    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = { type: 'update', name: '0.compress', value: false }
    const newState = dispatchNewHistoryConfig(action)
    expect(newState[0].compress).toEqual(false)
    expect(newState.errors).toBeUndefined()

    // remove test error
    delete testConfig.errors
  })
  test('check update action invalide', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = {
      type: 'update',
      name: '0.compress',
      value: 0,
      validity: 'Value should be boolean',
    }
    const newState = dispatchNewHistoryConfig(action)
    expect(newState.errors['0.compress']).toEqual(
      'Value should be boolean',
    )
  })
  test('check deleteRow action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })

    const action = { type: 'deleteRow', name: '0' }
    const newState = dispatchNewHistoryConfig(action)
    expect(newState).toEqual([testConfig[1]])
  })
  test('check deleteRow action with previous error', () => {
    // add some previuous error to test error cleaning
    testConfig.errors = { 0: 'some previous error' }
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })

    const action = { type: 'deleteRow', name: '0' }
    const newState = dispatchNewHistoryConfig(action)
    const expectedState = utils.jsonCopy(testConfig)
    delete expectedState.errors
    expect(newState).toEqual([expectedState[1]])

    // remove test error
    delete testConfig.errors
  })
  test('check deleteAllRows action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = {
      type: 'deleteAllRows',
      name: '0.points',
    }
    const newState = dispatchNewHistoryConfig(action)
    const expectedState = utils.jsonCopy(testConfig)
    expectedState[0].points = []
    expect(newState).toEqual(expectedState)
  })
  test('check addRow action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = {
      type: 'addRow',
      name: '0.points',
      value: {},
    }
    const newState = dispatchNewHistoryConfig(action)
    const expectedState = utils.jsonCopy(testConfig)
    expectedState[0].points.push({})
    expect(newState).toEqual(expectedState)
  })
  test('check importPoints action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = {
      type: 'importPoints',
      name: '0.points',
      value: [{}],
    }
    const newState = dispatchNewHistoryConfig(action)
    const expectedState = utils.jsonCopy(testConfig)
    expectedState[0].points = [{}]
    expect(newState).toEqual(expectedState)
  })
  test('check unknown action', () => {
    act(() => {
      ReactDOM.render(
        <HistoryConfigProvider>
          <div />
        </HistoryConfigProvider>,
        container,
      )
    })
    const action = { type: 'unknown' }
    expect(() => dispatchNewHistoryConfig(action)).toThrow()
  })
})
