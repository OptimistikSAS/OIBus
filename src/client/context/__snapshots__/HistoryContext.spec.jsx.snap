// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryConfigProvider check HistoryConfigProvider 1`] = `
<div>
  <div />
</div>
`;

exports[`HistoryConfigProvider check initial data setup 1`] = `
<div>
  <div />
</div>
`;
